lib/src/compiler/parse.rs

```diff
@@ -1,14 +1,220 @@
-use crate::{ast::Program, compiler::Compiler};
+use crate::{
+    ast::{Instr, LangValue, Program, StackIdentifier},
+    compiler::Compiler,
+    error::ParseError,
+};
 use failure::Fallible;
+use nom::{
+    branch::alt,
+    bytes::complete::tag,
+    character::complete::{char, digit1, multispace0},
+    combinator::map_res,
+    multi::{many0, many1},
+    sequence::{delimited, preceded, terminated},
+    IResult,
+};
 use std::io::Read;
 
+fn parse_read(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Read")(input)?;
+    Ok((input, Instr::Read))
+}
+
+fn parse_write(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Write")(input)?;
+    Ok((input, Instr::Write))
+}
+
+fn parse_set(input: &str) -> IResult<&str, Instr> {
+    let (input, _) = tag("Set")(input)?;
+    let (input, val) = preceded(
+        multispace0,
+        map_res(digit1, |s: &str| s.parse::<LangValue>()),
+    )(input)?;
```

Comment:
Can you add a comment on each one of these with a little example of what it's parsing? idk what all this `preceded` and `multispace0` mumbo jumbo is.

Reply:
On it